### PR TITLE
Add STM32L433 and L443 support

### PR DESCRIPTION
**Pull request by Xelus22**
## Description

Add STM32L433 and STM32L443 support.
Currently untested on hardware. Will change to a proper PR once tested.

Edit: Tested and working with i2c and spi. However PWM support is not yet tested

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Adds STM32L433 and STM32L443 support

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: lib/python/qmk/constants.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 MAX_KEYBOARD_SUBFOLDERS = 5
 
 # Supported processor types
-CHIBIOS_PROCESSORS = 'cortex-m0', 'cortex-m0plus', 'cortex-m3', 'cortex-m4', 'MKL26Z64', 'MK20DX128', 'MK20DX256', 'STM32F042', 'STM32F072', 'STM32F103', 'STM32F303', 'STM32F401', 'STM32F411', 'STM32G431', 'STM32G474'
+CHIBIOS_PROCESSORS = 'cortex-m0', 'cortex-m0plus', 'cortex-m3', 'cortex-m4', 'MKL26Z64', 'MK20DX128', 'MK20DX256', 'STM32F042', 'STM32F072', 'STM32F103', 'STM32F303', 'STM32F401', 'STM32F411', 'STM32G431', 'STM32G474', 'STM32L433'
```
Comment (tzarc) [inline lib/python/qmk/constants.py:13]: Also `qmk_firmware/data/schemas/keyboard.jsonschema`

File: platforms/chibios/GENERIC_STM32_L433XC/configs/mcuconf.h [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+#define STM32_ICU_USE_TIM15                 FALSE
+#define STM32_ICU_USE_TIM16                 FALSE
+
+/*
+ * PWM driver system settings.
+ */
+#define STM32_PWM_USE_ADVANCED              FALSE
+#define STM32_PWM_USE_TIM1                  FALSE
+#define STM32_PWM_USE_TIM2                  FALSE
+#define STM32_PWM_USE_TIM15                 TRUE
```
Comment (tzarc) [inline platforms/chibios/GENERIC_STM32_L433XC/configs/mcuconf.h:186]: Please disable all peripherals by default.

File: lib/python/qmk/constants.py (hunk 1 above)
Comment (Xelus22) [reply lib/python/qmk/constants.py:13]: Yep its added already

File: platforms/chibios/GENERIC_STM32_L433XC/configs/mcuconf.h (hunk 2 above)
Comment (Xelus22) [reply platforms/chibios/GENERIC_STM32_L433XC/configs/mcuconf.h:186]: Oops, will update.

File: lib/python/qmk/constants.py (hunk 1 above)
Comment (tzarc) [reply lib/python/qmk/constants.py:13]: I'm blind, don't mind me.

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by donicrosby** [comment]
Hi I made #12593 would you like some help with this? The PCB I designed uses this MCU, so if you would like an extra set of hands let me know.

**Comment by Xelus22** [comment]
Hey @donicrosby , the only thing left to test now is the PWM support for both backlight and the WS2812 PWM driver for RGBLIGHT/RGBMATRIX.

**Comment by tzarc** [comment]
> 
> 
> Hey @donicrosby , the only thing left to test now is the PWM support for both backlight and the WS2812 PWM driver for RGBLIGHT/RGBMATRIX.

At this point I'd be fine with tackling that as a followup PR.

File: lib/python/qmk/constants.py (hunk 1 above)
Comment (tzarc) [reply lib/python/qmk/constants.py:13]: Should add L443, too.

File: data/schemas/keyboard.jsonschema [hunk 3]
```diff
@@ -25,7 +25,7 @@
         },
         "processor": {
             "type": "string",
-            "enum": ["cortex-m0", "cortex-m0plus", "cortex-m3", "cortex-m4", "MKL26Z64", "MK20DX128", "MK20DX256", "STM32F042", "STM32F072", "STM32F103", "STM32F303", "STM32F401", "STM32F411", "STM32G431", "STM32G474", "atmega16u2", "atmega32u2", "atmega16u4", "atmega32u4", "at90usb162", "at90usb646", "at90usb647", "at90usb1286", "at90usb1287", "atmega32a", "atmega328p", "atmega328", "attiny85", "unknown"]
+            "enum": ["cortex-m0", "cortex-m0plus", "cortex-m3", "cortex-m4", "MKL26Z64", "MK20DX128", "MK20DX256", "STM32F042", "STM32F072", "STM32F103", "STM32F303", "STM32F401", "STM32F411", "STM32G431", "STM32G474", "STM32L433", "atmega16u2", "atmega32u2", "atmega16u4", "atmega32u4", "at90usb162", "at90usb646", "at90usb647", "at90usb1286", "at90usb1287", "atmega32a", "atmega328p", "atmega328", "attiny85", "unknown"]
```
Comment (tzarc) [inline data/schemas/keyboard.jsonschema:28]: L443 here too.

File: platforms/chibios/GENERIC_STM32_L433XC/configs/config.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* Address for jumping to bootloader on STM32 chips. */
+/* It is chip dependent, the correct number can be looked up here (page 175):
+ * http://www.st.com/web/en/resource/technical/document/application_note/CD00167594.pdf
+ * This also requires a patch to chibios:
+ *  <tmk_dir>/tmk_core/tool/chibios/ch-bootloader-jump.patch
+ */
```
Comment (tzarc) [inline platforms/chibios/GENERIC_STM32_L433XC/configs/config.h:22]:
```suggestion
/* It is chip dependent, the correct number can be looked up by checking against ST's application note AN2606.
 */
```

File: docs/compatible_microcontrollers.md [hunk 5]
```diff
@@ -30,6 +30,7 @@ You can also use any ARM chip with USB that [ChibiOS](https://www.chibios.org) s
  * [STM32F411](https://www.st.com/en/microcontrollers-microprocessors/stm32f411.html)
  * [STM32G431](https://www.st.com/en/microcontrollers-microprocessors/stm32g4x1.html)
  * [STM32G474](https://www.st.com/en/microcontrollers-microprocessors/stm32g4x4.html)
+ * [STM32L433](https://www.st.com/en/microcontrollers-microprocessors/stm32l4x3.html)
```
Comment (tzarc) [inline docs/compatible_microcontrollers.md:33]:
```suggestion
 * [STM32L433](https://www.st.com/en/microcontrollers-microprocessors/stm32l4x3.html)
 * [STM32L443](https://www.st.com/en/microcontrollers-microprocessors/stm32l4x3.html)
```

File: docs/ja/compatible_microcontrollers.md [hunk 6]
```diff
@@ -35,6 +35,7 @@ QMK は十分な容量のフラッシュメモリを備えた USB 対応 AVR ま
 * [STM32F411](https://www.st.com/en/microcontrollers-microprocessors/stm32f411.html)
 * [STM32G431](https://www.st.com/en/microcontrollers-microprocessors/stm32g4x1.html)
 * [STM32G474](https://www.st.com/en/microcontrollers-microprocessors/stm32g4x4.html)
+* [STM32L433](https://www.st.com/en/microcontrollers-microprocessors/stm32l4x3.html)
```
Comment (tzarc) [inline docs/ja/compatible_microcontrollers.md:38]:
```suggestion
* [STM32L433](https://www.st.com/en/microcontrollers-microprocessors/stm32l4x3.html)
* [STM32L443](https://www.st.com/en/microcontrollers-microprocessors/stm32l4x3.html)
```

File: platforms/chibios/GENERIC_STM32_L433XC/configs/board.h [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+#pragma once
+
+#include_next "board.h"
+
+#undef STM32L432xx
+
+// Use L443xx define instead of L433xx since ChibiOS defines
+// L433xx together with L432x, which does not activate the 
+// GPIO H ports and does not enable I2C2 and SPI2
+// Until the bug is fixed in ChibiOS, this should not change
```
Comment (tzarc) [inline platforms/chibios/GENERIC_STM32_L433XC/configs/board.h:25]:
```suggestion
// Pretend that we're an L443xx as the ChibiOS definitions for L432 don't enable GPIOH, I2C2, or SPI2.
// Until ChibiOS upstream is fixed, this should be kept at L443, as nothing in QMK currently utilises
// the crypto peripheral on the L443.
```

**Review by tzarc — CHANGES_REQUESTED** [review]
Pretty sure we're going to have to go to `develop` for this.

**Comment by tzarc** [comment]
Apologies, got a couple of minor conflicts after changing target branch.

File: quantum/mcu_selection.mk [hunk 8]
```diff
@@ -396,6 +396,41 @@ ifneq ($(findstring STM32G474, $(MCU)),)
...
+  # <keyboard_dir>/boards/, or drivers/boards/
+  BOARD ?= GENERIC_STM32_L433XC
+
+  PLATFORM_NAME ?= platform_l432
+
+  USE_FPU ?= yes
+
+  # Options to pass to dfu-util when flashing
+  DFU_ARGS ?= -d 0483:DF11 -a 0 -s 0x08000000:leave
+  DFU_SUFFIX_ARGS ?= -v 0483 -p DF11
```
Comment (tzarc) [inline quantum/mcu_selection.mk:431]:
```suggestion
  DFU_SUFFIX_ARGS ?= -v 0483 -p DF11

  # UF2 settings
  UF2_FAMILY ?= STM32L4
```

**Review by tzarc — CHANGES_REQUESTED** [review]
Just a small change -- you've also got some conflicts due to the F446 support getting added to `develop`.
Rest looks good!

**Review by tzarc — APPROVED** [review]

**Pull request merged**